pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,229 @@
+                             callback: ApplicationHandler,
+                             ack_queue: 'asyncio.Queue[str]',
+                             metrics_client: Optional[MetricsAgent]
+                             ) -> None:
+        try:
+            start = time.perf_counter()
+            await callback(message)
+            await ack_queue.put(message.ack_id)
+            if metrics_client:
+                metrics_client.increment('pubsub.consumer.succeeded')
+                metrics_client.histogram('pubsub.consumer.latency.runtime',
+                                         time.perf_counter() - start)
+        except Exception:
+            log.exception('Application callback raised an exception')
+            if metrics_client:
+                metrics_client.increment('pubsub.consumer.failed')
+        finally:
+            semaphore.release()
+
+    async def consumer(
+            message_queue: MessageQueue,
+            callback: ApplicationHandler,
+            ack_queue: 'asyncio.Queue[str]',
+            ack_deadline_cache: AckDeadlineCache,
+            consumer_pool_size: int,
+            metrics_client: Optional[MetricsAgent]) -> None:
+        semaphore = asyncio.Semaphore(consumer_pool_size)
+        while True:
+            await semaphore.acquire()
+
+            message, pulled_at = await message_queue.get()
+
+            ack_deadline = await ack_deadline_cache.get()
+            if (time.perf_counter() - pulled_at) >= ack_deadline:
+                if metrics_client:
+                    metrics_client.increment('pubsub.consumer.failfast')
+                message_queue.task_done()
+                semaphore.release()
+                continue
+
+            if metrics_client:
+                metrics_client.histogram(
+                    'pubsub.consumer.latency.receive',
+                    time.time() - message.publish_time.timestamp())
+
+            asyncio.ensure_future(_fire_callback(
+                semaphore,
+                message,
+                callback,
+                ack_queue,
+                metrics_client,
+            ))
+            message_queue.task_done()
+
+    async def producer(
+            subscription: str,
+            message_queue: MessageQueue,
+            subscriber_client: 'SubscriberClient',
+            max_messages: int,
+            metrics_client: Optional[MetricsAgent]
+    ) -> None:
+        while True:
+            try:
+                new_messages = await subscriber_client.pull(
+                    subscription=subscription, max_messages=max_messages)
+            except asyncio.TimeoutError:
+                continue
+
+            pulled_at = time.perf_counter()
+            for m in new_messages:
+                await message_queue.put((m, pulled_at))
+
+            if metrics_client:
+                metrics_client.histogram(
+                    'pubsub.producer.batch', len(new_messages))
+
+            await message_queue.join()
```

Comment:
Intermediate results. Tested with `egress`, i.e. very fast consumer.
In high load scenarios there's no difference in performance or CPU usage.
Under small load not having `message_queue.join()`  generate occasional spikes in receive latency:
![image](https://user-images.githubusercontent.com/62574824/101955303-6a62a800-3bb2-11eb-982d-83d642f4fb76.png)
Not sure why, tho.

I'll be testing against ctm next

Reply:
Hmm, `ctm` looks alright. Now when I think about it, it looks like `egress` profile _is_ the worst case scenario for this 🤔  When `callback` is fast and asynchronous, there are plenty of time slots for consumer to fetch too many messages. But this seems to be not the case for the current design.  I'll go ahead and implement single message queue variant and test it

Reply:
Alright, bottomline:
0. current solution
1. no `message_queue.join()` -> same performance
2. no `message_queue.join()` + single queue + arbitrary number of consumer workers -> same performance, higher memory consumption
3. no `message_queue.join()` + single queue + single consumer with arbitrary large semaphore-base tasks pool -> poor performance, higher memory consumption

We could go with either 0 or 1.  Personally I lean toward going with current solution, since it's exactly what was tested throughout the whole process and it does not maintain an implicit pool of messages in memory.

Please let me know what you think. With this decision the PR should be unblocked for final polishing.

Reply:
Awesome results, thanks! Option 0 seems reasonable to me, but I'm glad we tested the other options just in case. The added memory consumption seems a bit odd to me, so I might play around on personal time to figure out what's up here, but we can always make changes as a future optimization rather than blocking this PR